File: frontend/src/app/contexts/journal/application/service/log-weight-measurement.service.ts
import type {
  LogWeightMeasurementCommand,
  LogWeightMeasurementResult,
  LogWeightMeasurementResultData,
  LogWeightMeasurementUseCase,
} from '../port/in/log-weight-measurement.use-case';
import type {
  LogWeightMeasurementInstruction,
  LogWeightMeasurementOutcomeData,
  WeightMeasurementStore,
} from '../port/out/weight-measurement.store';
import { calendarDate } from '../../../../shared/api/shared.calendar';
import { WeightMeasurement } from '../../domain/weight-measurement';
import { Weight } from '../../domain/weight';

export class LogWeightMeasurementService implements LogWeightMeasurementUseCase {
  constructor(private readonly store: WeightMeasurementStore) {}

  async log(command: LogWeightMeasurementCommand): Promise<LogWeightMeasurementResult> {
    if (!command) {
      throw new Error('command must not be null');
    }

    const domain = commandToDomain(command);

    const instruction = toInstruction(domain);
    const outcome = await this.store.log(instruction);

    if (outcome.kind === 'failed') {
      throw new Error('Failed to log weight measurement: ' + outcome.outcomeData.errorMessage);
    }

    const domainResponse = outcomeDataToDomain(outcome.outcomeData);

    return {
      kind: 'logged',
      resultData: toResultData(domainResponse),
    };
  }
}

function commandToDomain(command: LogWeightMeasurementCommand): WeightMeasurement {
  return WeightMeasurement.create(calendarDate(command.date), Weight.of(command.weightInKg));
}

function outcomeDataToDomain(command: LogWeightMeasurementOutcomeData): WeightMeasurement {
  return WeightMeasurement.create(calendarDate(command.date), Weight.of(command.weightInKg));
}

function toInstruction(domain: WeightMeasurement): LogWeightMeasurementInstruction {
  return {
    date: domain.calendarDate,
    weightInKg: domain.weightInKilograms(),
  };
}

function toResultData(domain: WeightMeasurement): LogWeightMeasurementResultData {
  return {
    date: domain.calendarDate,
    weightInKg: domain.weightInKilograms(),
  };
}

File: backend/src/main/java/fr/sylvainjanet/tracker/journal/application/service/LogWeightMeasurementService.java
package fr.sylvainjanet.tracker.journal.application.service;

import static fr.sylvainjanet.tracker.journal.domain.builder.WeightMeasurementBuilder.aWeightMeasurement;

import fr.sylvainjanet.tracker.journal.application.port.in.dtos.command.LogWeightMeasurementCommand;
import fr.sylvainjanet.tracker.journal.application.port.in.dtos.result.LogWeightMeasurementResult;
import fr.sylvainjanet.tracker.journal.application.port.in.usecase.LogWeightMeasurementUseCase;
import fr.sylvainjanet.tracker.journal.application.port.out.dtos.instruction.LogWeightMeasurementInstruction;
import fr.sylvainjanet.tracker.journal.application.port.out.dtos.outcome.LogWeightMeasurementOutcome;
import fr.sylvainjanet.tracker.journal.application.port.out.gateway.store.WeightMeasurementStore;
import fr.sylvainjanet.tracker.journal.domain.WeightMeasurement;
import java.util.Objects;

public final class LogWeightMeasurementService implements LogWeightMeasurementUseCase {

    private final WeightMeasurementStore store;

    public LogWeightMeasurementService(WeightMeasurementStore store) {
        this.store = Objects.requireNonNull(store, "store must not be null");
    }

    @Override
    public LogWeightMeasurementResult log(LogWeightMeasurementCommand command) {
        Objects.requireNonNull(command, "command must not be null");

        WeightMeasurement domainCommand = commandToDomain(command);

        LogWeightMeasurementOutcome outcome = store.log(toInstruction(domainCommand));

        WeightMeasurement domainOutcome = outcomeToDomain(outcome);

        return toResult(domainOutcome);
    }

    private WeightMeasurement commandToDomain(LogWeightMeasurementCommand command) {
        return aWeightMeasurement()
                .withDate(command.date())
                .withWeightInKg(command.weightInKg())
                .build();
    }

    private LogWeightMeasurementInstruction toInstruction(WeightMeasurement weightMeasurement) {
        return new LogWeightMeasurementInstruction(
                weightMeasurement.date(), weightMeasurement.weightInKilograms());
    }

    private LogWeightMeasurementResult toResult(WeightMeasurement weightMeasurement) {
        return new LogWeightMeasurementResult(
                weightMeasurement.date(), weightMeasurement.weightInKilograms());
    }

    private WeightMeasurement outcomeToDomain(LogWeightMeasurementOutcome outcome) {
        return aWeightMeasurement()
                .withDate(outcome.date())
                .withWeightInKg(outcome.weightInKg())
                .build();
    }
}
